src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -1722,6 +1861,74 @@ protected void move(File source, File target, String entity) throws IOException,
         }
     }
 
+    /**
+     * Attempt to delete an object based on the given content identifier (cid). If the object has
+     * pids that references it and/or a cid refs file exists, the object will not be deleted.
+     *
+     * @param cid Content identifier
+     * @throws IOException              If an issue arises during deletion of object
+     * @throws NoSuchAlgorithmException Incompatible algorithm used to find relative path to cid
+     * @throws InterruptedException     Issue with synchronization of cid deletion
+     */
+    protected void deleteObjectByCid(String cid)
+        throws IOException, NoSuchAlgorithmException, InterruptedException {
+        Path absCidRefsPath = getExpectedPath(cid, "refs", HashStoreIdTypes.cid.getName());
+        if (Files.exists(absCidRefsPath)) {
+            String warnMsg = "FileHashStore - deleteObjectByCid: cid refs file still contains"
+                + " references, skipping deletion.";
+            logFileHashStore.warn(warnMsg);
+            // The cid is referenced by pids, do not delete.
+
+        } else {
+            // Get permanent address of the actual cid
+            String objRelativePath = FileHashStoreUtility.getHierarchicalPathString(
+                DIRECTORY_DEPTH, DIRECTORY_WIDTH, cid
+            );
+            Path expectedRealPath = OBJECT_STORE_DIRECTORY.resolve(objRelativePath);
+
+            // Minimize the amount of time the cid is locked
+            synchronized (referenceLockedCids) {
```

Comment:
This `synchronized` block is not located in the `try-finally` block which will remove the cids. So it can create a dead lock since a cid can never be removed from the array.
You need to check all `synchronized` sections in this class.

Reply:
Thank you for catching this! I have fixed this, and also fixed a bug uncovered as I was double checking the class.